proposed/logging-generator.md

```diff
@@ -0,0 +1,308 @@
+        }
+    }
+}
+```
+
+An argument against `LoggerMessage.Define` APIs through community has been that it is hard to use, and there is a maintainance burden that hurts usability.
+
+## Requirements
+
+### Goals
+
+The number one concern raised with logging seems to be, in terms of usability, on finding a way to enforce strong typing on logging. In other words, it would be great to come up with a strong convention for all developers in a project to follow a specific logging-only template.
+
+### Non-Goals
+
+Also it's been raised that `LoggerMessage.Define` approach triggers boxing of data, and we wanted to understand if there are ways to improve this with a new and improved design.
+
+
+## Design
+
+`LoggerMessageAttribute` As shown in the examples, is used by the developer to trigger source generation:
+
+```c#
+namespace Microsoft.Extensions.Logging
+{
+    [AttributeUsage(AttributeTargets.Method)]
+    public sealed partial class LoggerMessageAttribute : Attribute
+    {
+        public LoggerMessageAttribute(int eventId, LogLevel level, string? message = null);
+        public LoggerMessageAttribute(int eventId, string? message = null);
+        public int EventId { get; }
+        public string? EventName { get; set; }
+        public LogLevel? Level { get; }
+        public string? Message { get; }
+    }
+}
+```
+
+## Usage Examples: 
+
+The developer could use a newly proposed `LoggerMessageAttribute` API in order to trigger code generation:
+
+```csharp
+public partial class LoggingSample3
+{
+    private readonly ILogger _logger;
+
+    public LoggingSample3(ILogger logger)
+    {
+        _logger = logger;
+    }
+
+    [LoggerMessage(0, LogLevel.Information, "Hello {name}")]
+    public partial void LogName(string name);
+}
+```
+
+Where the implementation for `LogName` would be completed by the source generator. 
+
+## More Scenarios
+
+#### **LoggingSample4:** with static class Log
+
+```csharp
+public static partial class Log
+{
+    [LoggerMessage(0, LogLevel.Information, "Hello `{name}`")]
+    public static partial void LogName(this ILogger logger, string name);
+}
+
+public class LoggingSample4
+{
+    private ILogger _logger;
+
+    public LoggingSample4(ILogger logger)
+    {
+        _logger = logger;
+    }
+
+    public void LogName(string name)
+    {
+        _logger.LogName(name);
+    }
+}
+```
+
+#### **LoggingSample5:** Other miscellaneous logging 
+
+```c#
+public partial class LoggingSample3
+{
+    private readonly ILogger _logger;
+
+    public LoggingSample3(ILogger logger)
+    {
+        _logger = logger;
+    }
+
+    [LoggerMessage(1, LogLevel.Trace)]
+    public partial void LogEmptyMessage();
+
+    [LoggerMessage(9, LogLevel.Trace, "Fixed message", EventName = "CustomEventName")]
+    public partial void LogWithCustomEventName();
+
+    [LoggerMessage(10, "Welcome to {city} {province}!")]
+    public partial void LogWithDynamicLogLevel(string city, LogLevel level, string province);
```

Comment:
My first thought when I saw this attribute was that this used some default log level (I didn't notice the parameter).  Do you think this may confuse customers? Is logging with a dynamic level common?

Reply:
Who changes log level per call @geeknoid?

Reply:
The log level is not normally changed per call, it's based on config.

The specific case is that we are building a generic facility to log HTTP requests. Someone wanting to use our component gets to decide the log level our log state comes out as. The first implementation of that code needed to declare different log methods, all with the same parameters, except for the log level. And then at runtime, it needed  switch statement to decide which version of the log method to invoke. Basically, it was quite lame. Making this small extension to the logger generator to make the log level available dynamically eliminated all that crazy redundancy.